src/eckit/config/Configuration.h

```diff
@@ -130,6 +131,70 @@ class Configuration : public Parametrisation {
 
     virtual void hash(eckit::Hash&) const;
 
+    // -- Introspection methods
+
+    bool isSubConfiguration(const std::string& name) const;
```

Comment:
List of methods that ideally should follow either eckit::Value variants, or <type_traits> (and maybe eckit::Value should follow that too). I only find the extensive list of isType... and the template isConvertible<Type> to be inconsistent (both should maybe be template, and leverage <type_traits>? not sure)

Reply:
Could you please show an example API of what you have in mind?

Reply:
I mean a choice between

1. comprehensively to the class accessors introspect the eckit::Configuration types, as
instead of (isSubConfiguration is special, so ommited):
`
    bool isIntegral(const std::string&) const;
    bool isBoolean(const std::string&) const;
    bool isFloatingPoint(const std::string&) const;
    bool isString(const std::string&) const;
    bool isList(const std::string&) const;
    bool isSubConfigurationList(const std::string&) const;
    bool isIntegralList(const std::string&) const;
    bool isBooleanList(const std::string&) const;
    bool isFloatingPointList(const std::string&) const;
    bool isStringList(const std::string&) const;`
    do
    `
    bool isBool(const std::string&) const;
    bool isInt(const std::string&) const;
    bool isLong(const std::string&) const;
    bool isUnsigned(const std::string&) const;
    bool isInt32(const std::string&) const;
    bool isInt64(const std::string&) const;
    bool isFloat(const std::string&) const;
    bool isDouble(const std::string&) const;
    bool isString(const std::string&) const;
    bool isIntVector(const std::string&) const;
    bool isLongVector(const std::string&) const;
    bool isUnsignedVector(const std::string&) const;
    bool isInt32Vector(const std::string&) const;
    bool isInt64Vector(const std::string&) const;
    bool isFloatVector(const std::string&) const;
    bool isDoubleVector(const std::string&) const;
    bool isStringVector(const std::string&) const;`
2. or, comprehensively (to eckit::Value, because this is what implements conversion) introspect the eckit::Configuration types, as instead of (same as above), do
`    bool isNumber(const std::string&) const;
    bool isBool(const std::string&) const;
    bool isDouble(const std::string&) const;
    bool isString(const std::string&) const;
    bool isNil(const std::string&) const;
    bool isList(const std::string&) const;
    bool isMap(const std::string&) const;
    bool isDate(const std::string&) const;
    bool isTime(const std::string&) const;
    bool isDateTime(const std::string&) const;
    bool isOrderedMap(const std::string&) const;`

In the end I'm only mentioning consistency so logical patterns emerge. I would think this is a good time to re-think some of the classes purposes as some inconsistency is of course already present -- otherwise we wouldn't need to discuss! :-)

Reply:
I see what you mean.

eckit::Value has "isNumber" and "isDouble" whereas I opted for "isIntegral" and "isFloatingPoint" which resembles the STL type_traits naming more closely. I'm OK to use eckit::Value nomenclature instead if that is  a more expected naming.

I am also OK to use the current Configuration "get<Type>" naming for "is<Type>" or "has<Type>".
Fact is that "getFloat" and "getDouble" both work. Same with "getInt" and "getLong".
Would you then say that "isFloat" and "isDouble" should both return "true" for an eckit::Value that encodes a double?
That is why I had opted for the "isConvertible<T>" in the end as an extra API on top of "isFloatingPoint".

Please let me know what you prefer, also @tlmquintino and @simondsmart and/or others , please provide input here.

Reply:
I completely agree! I'm only trying to limit the amount of variations on these "families of methods" across the classes (I count 3 variants before the PR?). An idea for the future would be to add a isConvertible to eckit::Value instead (avoiding some if's, the necessary hardcode would fall in the mythical "right place" (tm)) and _Configuration_ delegating to that.

I think we're not really discussing code but architecture, as the PR is obviously excellent (approved!), we only need to find if it fits the classes intention.

Reply:
one small comment. We explicitly do not want a method to expose the internal eckit::Value root.

Reply:
Already had discussions with @simondsmart half a year ago.

One the one side we do not want to expose `eckit::Value`, on the otherside we have no mechanisms to efficiently copy a value from one map to another.
Inspecting a type is very cumbersome. In multio we have some code that runs a lot of `if (isList()) else if (isMap()) else if(isNumber()) ...`. Which is super inefficient, many unnecessary function calls and checks.

Given that we the list of types is already fixed, it would be more useful to 
* either create a `enum ValueType` listing all the types and a virtual function `ValueType getType() const` to retrieve the type and properly switch on it 
* or to implement a visitor with a double-disptach, on which a properly typed overload is called with a reference to the value.. This would allow doing every typed specialization we want behind the interface boundary with 2 function calls.

Then we can use `std::is_convertible_v` (https://en.cppreference.com/w/cpp/types/is_convertible) or `eckit::IsTranslatable` in combination with `eckit::translate`.

The `isConvertible` call is likely to be followed by further inspection calls or eventually by a `get` call... which will quickly sum up in a lot of more calls... Instead with a visitor we are already describing what we are going to do.

With C++17 and lambdas creating a Visitor is also not that cumbersome  anymore.
We can always  create an overload set and generate most of the visitor.

E.g.

```
value.visit(Overloaded{
[&](std::string& str) {},
...
[&](auto& v){
// Anything else
}])
```

or

```
value.visit([&](auto& v){
  using T = std::decay_t<decltype(v)>;
  if constexpr (std::is_integral_v<T>) {
  } else if constexpr (...){
  }
})
```

We can even pretend to have arbitrary return values behind the interface by using a mutable Visitor that assigns to an `std::optional<RetType>` etc.

I don't think it is feasible to more and more functions on top of the interface to get the "type problem" solved.
With the way modern C++ is taking, I would even avoid hiding everything behind a dynamic interface unless there's a real need to and otherwise use all the template & type mechanisms C++ offers.

All these compile time features c++ offers are clearly things that allows C++ to compile & optimize more and distinguishes it from most other languages.


With C++20 concepts & templated lambdas this will eventually also be readable, with the same logic behind.

Reply:
After reading this comment I was wondering how multio managed to do this introspection without this PR changes.
It seems that you're breaking the contract and are extracting the eckit::Value from the LocalConfiguration. This should not have been allowed. e.g. here:
```c++
eckit::Value getInputGrid(const eckit::LocalConfiguration& cfg, message::Metadata& md) {
    if (md.has("atlas-grid-kind")) {  // metadata has always precedence
        // TODO: name is bad on purpose (no software support this at the moment)
        return eckit::Value{md.getSubConfiguration("atlas-grid-kind").get()};
    }

    if (cfg.has("input")) {  // configuration file is second option
        return eckit::Value{cfg.getSubConfiguration("input").get()};
    }

    throw eckit::SeriousBug("action-interpolate :: Unable to identify input grid", Here());
}
```
And then down the line `eckit::Value` is used to do the introspection (perhaps for another occurrence of this pattern).

Reply:
Well, I was done already somewhere else. Think even in atlas? And without that there is no way to copy an entry with arbitrary type from one configuration to another...

But the points I want to make:

* We already deal with a fixed set of types. So I think we should make that more explicit to use the other features & power of C++ and avoid the definitions of a lot of virtual functions....

* `eckit::Configuration` etc. are actually a map. If we do not expose an interfacing value type. All access and operations have to be made through a lookup first and then on the type. With all the introspecive operations we will repeatedly lookup the same key just to operate on the value. 
  The pure C++ way would probably be to perform a lookup and return an iterator - that's not possible for an interface. However, that's more reason to have a wrapper type like `eckit::Value`.